src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1319,6 +1355,201 @@ CreateCitusTable(Oid relationId, CitusTableType tableType,
+	if (tableType == SINGLE_SHARD_DISTRIBUTED)
+	{
+		uint32 targetNodeId = SingleShardTableColocationNodeId(colocationId);
+		if (targetNodeId != CoordinatorNodeIfAddedAsWorkerOrError()->nodeId)
+		{
+			bool missingOk = false;
+			WorkerNode *targetNode = FindNodeWithNodeId(targetNodeId, missingOk);
+			targetNodeList = list_make1(targetNode);
+
+			shouldDropLocalPlacement = true;
+		}
+	}
+	else if (tableType == REFERENCE_TABLE)
+	{
+		targetNodeList = ActivePrimaryNonCoordinatorNodeList(ShareLock);
+		targetNodeList = SortList(targetNodeList, CompareWorkerNodes);
+	}
+
+	bool autoConverted = false;
+	UpdateNoneDistTableMetadataGlobally(
+		relationId, citusTableParams.replicationModel,
+		colocationId, autoConverted);
+
+	/* recreate the shard placement on workers and insert into pg_dist_placement globally */
+	if (list_length(targetNodeList) > 0)
+	{
+		NoneDistTableReplicateCoordinatorPlacement(relationId, targetNodeList);
+	}
+
+	if (shouldDropLocalPlacement)
+	{
+		/*
+		 * We don't yet drop the local placement before handling partitions.
+		 * Otherewise, local shard placements of the partitions will be gone
+		 * before we create them on workers.
+		 *
+		 * However, we need to delete the related entry from pg_dist_placement
+		 * before distributing partitions (if any) because we need a sane metadata
+		 * state before doing so.
+		 */
+		NoneDistTableDeleteCoordinatorPlacement(relationId);
+	}
+
+	/* if this table is partitioned table, distribute its partitions too */
+	if (PartitionedTable(relationId))
+	{
+		/* right now we don't allow partitioned reference tables */
+		Assert(tableType == SINGLE_SHARD_DISTRIBUTED);
+
+		List *partitionList = PartitionList(relationId);
+
+		char *parentRelationName = generate_qualified_relation_name(relationId);
+
+		/*
+		 * When there are many partitions, each call to
+		 * ConvertCitusLocalTableToTableType accumulates used memory.
+		 * Create and free citus_per_partition_context for each call.
+		 */
+		MemoryContext citusPartitionContext =
+			AllocSetContextCreate(CurrentMemoryContext,
+								  "citus_per_partition_context",
+								  ALLOCSET_DEFAULT_SIZES);
+		MemoryContext oldContext = MemoryContextSwitchTo(citusPartitionContext);
+
+		Oid partitionRelationId = InvalidOid;
+		foreach_oid(partitionRelationId, partitionList)
+		{
+			MemoryContextReset(citusPartitionContext);
+
+			DistributedTableParams childDistributedTableParams = {
+				.colocationParam = {
+					.colocationParamType = COLOCATE_WITH_TABLE_LIKE_OPT,
+					.colocateWithTableName = parentRelationName,
+				},
+				.shardCount = distributedTableParams->shardCount,
+				.shardCountIsStrict = false,
+				.distributionColumnName = distributedTableParams->distributionColumnName,
+			};
+			ConvertCitusLocalTableToTableType(partitionRelationId, tableType,
+											  &childDistributedTableParams);
+		}
+
+		MemoryContextSwitchTo(oldContext);
+		MemoryContextDelete(citusPartitionContext);
+	}
+
+	if (shouldDropLocalPlacement)
```

Comment:
could be nice to have an extra comment on which cases this is trying to distinguish / why we need it